hazelcast/src/main/java/com/hazelcast/sql/impl/type/QueryDataTypeUtils.java

```diff
@@ -0,0 +1,261 @@
+                } else if (converter == InstantConverter.INSTANCE) {
+                    return QueryDataType.TIMESTAMP_WITH_TZ_INSTANT;
+                } else if (converter == OffsetDateTimeConverter.INSTANCE) {
+                    return QueryDataType.TIMESTAMP_WITH_TZ_OFFSET_DATE_TIME;
+                } else {
+                    assert converter == ZonedDateTimeConverter.INSTANCE;
+
+                    return QueryDataType.TIMESTAMP_WITH_TZ_ZONED_DATE_TIME;
+                }
+
+            case OBJECT:
+                return QueryDataType.OBJECT;
+
+            default:
+                assert typeFamily == QueryDataTypeFamily.LATE;
+
+                throw new IllegalArgumentException("Unexpected class: " + clazz);
+        }
+    }
+
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:ReturnCount", "checkstyle:MethodLength"})
+    public static QueryDataType resolveTypeForTypeFamily(QueryDataTypeFamily typeFamily) {
+        switch (typeFamily) {
+            case VARCHAR:
+                return QueryDataType.VARCHAR;
+
+            case BIT:
+                return QueryDataType.BIT;
+
+            case TINYINT:
+                return QueryDataType.TINYINT;
+
+            case SMALLINT:
+                return QueryDataType.SMALLINT;
+
+            case INT:
+                return QueryDataType.INT;
+
+            case BIGINT:
+                return QueryDataType.BIGINT;
+
+            case DECIMAL:
+                return QueryDataType.DECIMAL;
+
+            case REAL:
+                return QueryDataType.REAL;
+
+            case DOUBLE:
+                return QueryDataType.DOUBLE;
+
+            case DATE:
+                return QueryDataType.DATE;
+
+            case TIME:
+                return QueryDataType.TIME;
+
+            case TIMESTAMP:
+                return QueryDataType.TIMESTAMP;
+
+            case TIMESTAMP_WITH_TIMEZONE:
+                return QueryDataType.TIMESTAMP_WITH_TZ_OFFSET_DATE_TIME;
+
+            case OBJECT:
+                return QueryDataType.OBJECT;
+
+            default:
+                assert typeFamily == QueryDataTypeFamily.LATE;
+
+                throw new IllegalArgumentException("Unexpected type family: " + typeFamily);
+        }
+    }
+
+    /**
+     * Get integer type for the given precision.
+     *
+     * @param precision Precision.
+     * @return Type.
+     */
+    public static QueryDataType integerType(int precision) {
+        if (precision == 0) {
+            throw new IllegalArgumentException("Precision cannot be zero.");
+        }
+
+        if (precision == QueryDataType.PRECISION_UNLIMITED) {
+            return QueryDataType.DECIMAL;
+        } else if (precision <= QueryDataType.PRECISION_BIGINT) {
+            return INTEGER_TYPES[precision];
+        } else {
+            return QueryDataType.DECIMAL;
+        }
+    }
+
+    public static QueryDataType bigger(QueryDataType first, QueryDataType second) {
```

Comment:
Looks we need some documentation here or/and in the design document: "bigger" and "precedence is an integer number which defines context-dependent type resolution rules" don't sound that self-explanatory.

Isn't the meaning of "precedence"/"bigger" kind of inverted/confusing here? The type with the higher precedence value and higher precession is considered "bigger". Usually the lower precedence value means higher priority and when comparing things higher priority means bigger. Probably "widest" is a better name for this method. It's especially confusing seeing `LATE` (which is a special name for unknown/object, AFAIU) having the lowest precedence value among all types which implies it's the the most specific/restrictive type, so every other type is bigger/wider than the unknown type.

And also types having `PRECISION_UNLIMITED` are really confusing, `PRECISION_UNLIMITED` is actually the lowest precision possible according to `bigger`, probably a better name is something like `PRECISION_NONE/UNSPECIFIED/UNAPPLICABLE`.

Reply:
The only thing is needed is an ability to establish a total order between types. `bigger` function picks the type with higher priority. How the comparison is implemented internally, and whether we use ">" or "<" for that is implementation-defined. Most vendors don't provide any specific information on these implementation details. I used the same approach as in MS SQL, where a type with higher precedence is set for types with higher priority [1]. I do not see the value of inverting it, as this is merely a question of taste:
> the data type with the lower precedence is first converted to the data type with the higher precedence

Since `LATE` has the lowest value of the precedence, it is always the type with the lowest precedence.

`widest` doesn't work, because, e.g. DATE is not wider than TIME, it just has higher precedence. `bigger` has the same issue. I renamed it to `withHigherPrecedence`, it should add clarity.

`PRECISION_UNLIMITED` is applied to several types, such as REAL/DOUBLE/DECIMAL, where we have a virtually unlimited number of significant digits. A type with a higher number of significant digits wins, e.g. INT(12) has precedence over INT(11). Does it address your comment?

Regarding the documentation, I tried to rephrase it a bit, but it is vague still:
> Precedence is a comparable value which is used for type inference in expressions

I couldn't come with a better explanation in the context of this document because precedence is used differently in expressions. E.g. "INT(11) + SMALLINT(7) -> BIGINT(12)", but `INT(11) + VARCHAR -> DECIMAL(UNLIMITED)`. So exact usage of precedence should be documented in expressions design, not here.

[1] https://docs.microsoft.com/en-us/sql/t-sql/data-types/data-type-precedence-transact-sql?view=sql-server-ver15

Reply:
> PRECISION_UNLIMITED is applied to several types, such as REAL/DOUBLE/DECIMAL, where we have a virtually unlimited number of significant digits. A type with a higher number of significant digits wins, e.g. INT(12) has precedence over INT(11). 

At the same time PRECISION_UNLIMITED is defined as -1. If we have two types belonging to the same family (having the same precedence), the one having PRECISION_UNLIMITED would be considered having less precedence comparing to the other one having precision of say 10, while it should be the opposite (AFAIU). At least some kind of assertion should be introduced to avoid such erroneous ordering of types. Another option, is to define PRECISION_UNLIMITED as Integer.MAX_VALUE, so we can avoid the erroneous ordering described above. And even better is to introduce something like PRECISION_NONE for types like VARCHAR or OBJECT for which the concept of precision doesn't apply at all (-1 with assertion it never participates in precision comparison?).

I understand that with the current family/precedence/type table it's impossible to order the types incorrectly, but to understand that one needs to scan and match all the definitions from `QueryDataTypeFamily` and `QueryDataType`. Moreover, if one recklessly edits the family/precedence/type table, there is no protection from the incorrect ordering.

Reply:
Added explicit assertions on the current behavior. Let's keep it simple for now. I think that we will have to rethink the approach in the future when JDBC is ready. In this case every type will have to report some precision, and both `-1` or `Integer.MAX_VALUE` will be inapplicable, say, for `DECIMAL`. But at the same time, if we report a precision, we will have to ensure that follow it, i.e. trim values. Since we do not need that at the moment, let's keep the implementation as simple as possible.